### PR TITLE
PERF: building MultiIndex with categorical levels

### LINKED ISSUE #22044: Multi-index and CategoricalIndex performance
Hi,

Building a multi-index from a categorical index should be instantaneous: labels are codes and the corresponding level is a CategoricalIndex with the same N categories and codes [0 ... N-1] as intended in

https://github.com/pandas-dev/pandas/blob/edb71fda022c6a155717e7a25679040ee0476639/pandas/core/arrays/categorical.py#L2510

Unfortunately that call is painfully slow

```python
import pandas as pd
import numpy as np

pd.__version__
# 0.23.3

x = np.linspace(-1, 1, 1_000_000)
i = pd.Index(x)
j = i.astype('category')

%timeit pd.MultiIndex.from_arrays([j])
180 ms ± 912 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

The line above from `categorical.py` is the first bottleneck. It takes 130 ms in this example.
We could go around an replace that line with something along

```python
indexer = np.arange(len(values.categories), dtype=values.codes.dtype)
categories = pd.CategoricalIndex([], fastpath=True).set_categories(values.categories)
categories._data._codes = indexer
categories._cleanup()
```

On my machine, this takes ~2 ms

The 50 missing milliseconds are spent calling `_shallow_copy` in `MultiIndex._set_levels`. That copy may not be that shallow: looks like this boils down to yet another variation on the `CategoricalIndex`

```python
%timeit categories._shallow_copy()
# 46 ms ± 763 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit pd.CategoricalIndex(categories, categories.categories, categories.ordered)
# 47.2 ms ± 237 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Note however that

```python
%timeit pd.CategoricalIndex(categories)
# 3.69 µs ± 2.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

I would have liked to make PR but I don't fully understand the code and all that class overloading

### PR DESCRIPTION
**Pull request by 0x0L**
- [x] closes #22044
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

```python
df = pd.DataFrame({
    'a': np.arange(1_000_000, dtype=np.int32),
    'b': np.arange(1_000_000, dtype=np.int64),
    'c': np.arange(1_000_000, dtype=float),
}).astype({'a': 'category', 'b': 'category'})

%timeit df.set_index(['a', 'b'])
```
On my machine, this takes ~20ms. The current implem takes 140ms.

Any suggestion for tests ?

### CODE REVIEW

**Comment by WillAyd** [comment]
> Any suggestion for tests ?

For performance related items you can create a benchmark in `asv_bench/benchmarks` probably in the multiindex_object.py module for this (though categorical.py and ctors.py maybe work to dependent on how you build test)

**Comment by 0x0L** [comment]
@WillAyd 
I can't think of any possible downside in performance since this is just avoiding the rebuild of the codes (which are just 0..N-1 by definition) through lookup. 

I was thinking more in terms of possible regressions. I'm not sure what I should really test or worry for. But I couldn't find any test using a MultiIndex with at least a CategoricalIndex level...

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -2666,9 +2666,9 @@ def _factorize_from_iterable(values):
         raise TypeError("Input must be list-like")
 
     if is_categorical(values):
-        if isinstance(values, (ABCCategoricalIndex, ABCSeries)):
-            values = values._values
-        categories = CategoricalIndex(values.categories, dtype=values.dtype)
+        values = CategoricalIndex(values)
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: can you add a comment here on what is happening

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -514,6 +514,7 @@ Performance Improvements
 - Improved performance of :meth:`read_csv` by faster concatenating date columns without extra conversion to string for integer/float zero and float ``NaN``; by faster checking the string for the possibility of being a date (:issue:`25754`)
 - Improved performance of :attr:`IntervalIndex.is_unique` by removing conversion to ``MultiIndex`` (:issue:`24813`)
 - Restored performance of :meth:`DatetimeIndex.__iter__` by re-enabling specialized code path (:issue:`26702`)
+- Improved performance of :meth:`DataFrame.set_index` when using multiple indexes and at least one of them is a categorical object (:issue:`22044`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you add the asv as you have in the example. Also we might not have tests for using multiple cats in a MI, see if you can add one in pandas/tests/indexes/multi/test_constructor which match what the .set_index() is doing at a lower level.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @0x0L very nice!

**Pull request merged**